lotman/lotman_linux.go

```diff
@@ -314,9 +333,580 @@ func getFederationIssuer() (string, error) {
+
+    // Check if the value has a suffix
+    if len(value) > 1 {
+        suffix := strings.ToLower(string(value[len(value)-1]))
+        if multiplier, exists := suffixMultipliers[suffix]; exists {
+            number, err := strconv.ParseUint(value[:len(value)-1], 10, 64)
+            if err != nil {
+                return 0, err
+            }
+            return number * multiplier, nil
+        }
+    }
+
+    // If no suffix, treat as percentage
+    percentage, err := strconv.ParseFloat(strings.TrimSuffix(value, "%"), 64)
+    if err != nil {
+        return 0, err
+    }
+    return uint64((percentage / 100) * float64(totalDiskSpace)), nil
+}
+
+
+// Divide the remaining space among lots' dedicatedGB values -- we don't ever want to
+// dedicate more space than we have available, as indicated by the HWM of the cache. This is because
+// our model is that as long as a lot stays under its dedicated GB, its data is safe in the cache --
+// If the sum of each lot's dedicated GB exceeds the HWM, the cache may purge data without a single lot
+// exceeding it's quota. BAD!
+//
+// Opportunistic space can (and should) be overallocated, so unless explicitly set, each lot will have
+// dedicatedGB + opportunisticGB = HWM. This isn't maxed out to the total disk space, because otherwise
+// no lot could ever exceed its opportunistic storage and we'd lose some of the capabilities to reason about
+// how greedy the lot is.
+func divideRemainingSpace(lotMap *map[string]Lot, totalDiskSpaceB uint64) error {
+	hwmStr := param.Cache_HighWaterMark.GetString()
+	if hwmStr == "" {
+		return errors.New("high watermark is not set in the cache configuration")
+	}
+	hwm, err := convertWatermarkToBytes(hwmStr, totalDiskSpaceB)
+	if err != nil {
+		return errors.Wrap(err, "error converting high watermark to byte value for Lotman")
+	}
+	remainingToHwmB := hwm
+
+    // first iterate through all lots and subtract from our total space any amount
+    // that's already been allocated. Note which lots have an unset value, as we'll
+    // need to return to them.
+    returnToKeys := make([]string, 0, len(*lotMap))
+	for key, lot := range *lotMap {
+		if lot.LotName == "root" {
+			continue
+		}
+		if lot.MPA != nil && lot.MPA.DedicatedGB != nil  {
+			remainingToHwmB -= gigabytesToBytes(*lot.MPA.DedicatedGB)
+			if lot.MPA.OpportunisticGB == nil {
+				oGb := bytesToGigabytes(hwm) - *lot.MPA.DedicatedGB
+				lot.MPA.OpportunisticGB = &oGb
+			}
+		} else {
+			returnToKeys = append(returnToKeys, lot.LotName)
+		}
+		(*lotMap)[key] = lot
+	}
+
+	if len(returnToKeys) > 0 {
+		// now iterate through the lots that need space allocated and assign them the
+		// remaining space
+		spacePerLotRemainingB := remainingToHwmB / uint64(len(returnToKeys))
+		for _, key := range returnToKeys {
+			lot := (*lotMap)[key]
+			if lot.MPA == nil {
+				lot.MPA = &MPA{}
+			}
+			dGb := bytesToGigabytes(spacePerLotRemainingB)
+			oGb := bytesToGigabytes(hwm - spacePerLotRemainingB)
```

Comment:
if in the above case, remainingToHwmB is negative, then spacePerLotRemainingB is negative which would make oGB higher than the watermark in this case.